api.go

```diff
@@ -354,6 +354,38 @@ func NonFinal(fn any) Provider {
 	})
 }
 
+// AllowReturnFootgun annotates a provider (assumed to be a
+// wrapper/middleware) to say that it is okay for the wrapper to return
+// a type (usually error) that it does not in turn receive from
+// its call to inner(). This is only a footgun (and error) if a downstream
+// provider returns that type.
+//
+// For example, in the following chain, "footgun" can loose the error
+// from the return to callSomething()
+//
+//	 err := nject.Run("bad",
+//		nject.Provide("footgun", func(inner func()) error {
+//			if someCondition {
+//				return fmt.Errorf("some condition happened")
+//			}
+//			inner()
+//		}),
+//		func() error {
+//			return callSomething()
+//		},
+//	)
+//
+// Starting with version 2.0, nject rejects chains with footguns.
+func AllowReturnFootgun[T any](fn any) Provider {
```

Comment:
@dncohen suggestion: call it `Shadowing` instead. Also: have a runnable example.